### PR TITLE
pytest for visualization.py

### PR DESCRIPTION
**Pull request by tamakoshi2001**
**Checklist**
Thank you for contributing to QuTiP! Please make sure you have finished the following tasks before opening the PR.

- [x] Please read [Contributing to QuTiP Development](http://qutip.org/docs/latest/development/contributing.html)
- [x] Contributions to qutip should follow the [pep8 style](https://www.python.org/dev/peps/pep-0008/).
You can use [pycodestyle](http://pycodestyle.pycqa.org/en/latest/index.html) to check your code automatically
- [x] Please add tests to cover your changes if applicable.
- [x] If the behavior of the code has changed or new feature has been added, please also update the documentation in the `doc` folder, and the [notebook](https://github.com/qutip/qutip-tutorials). Feel free to ask if you are not sure.
- [x] Include the changelog in a file named: `doc/changes/<PR number>.<type>` 'type' can be one of the following: feature, bugfix, doc, removal, misc, or deprecation (see [here](http://qutip.org/docs/latest/development/contributing.html#changelog-generation) for more information).

Delete this checklist after you have completed all the tasks. If you have not finished them all, you can also open a [Draft Pull Request](https://github.blog/2019-02-14-introducing-draft-pull-requests/) to let the others know this on-going work and keep this checklist in the PR description.

**Description**
pytest for #2170

### CODE REVIEW

**Comment by tamakoshi2001** [comment]
Here is a sample pytest for functions in visualization.py

File: qutip/tests/test_visualization.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+     None, [1, 2, 3, 4, 5], "scaled", True, None, None, None, None),
+    (qutip.rand_dm(5), None, None, "scaled", True, None, None, None, None),
+    (qutip.rand_dm(5), None, None, "threshold", True, None, None, None, None),
+    (qutip.rand_dm(5), None, None, "phase", True, None, None, None, None),
+])
+def test_hinton(rho, x_basis, y_basis, color_style,
+                label_top, cmap, colorbar, fig, ax):
+    fig, ax = qutip.hinton(rho, x_basis, y_basis, color_style,
+                           label_top, cmap=cmap, colorbar=colorbar,
+                           fig=fig, ax=ax)
```
Comment (Ericgig) [inline qutip/tests/test_visualization.py:27]: - It's better not to call qutip function in the parametrization step, if there is a bug in that function, it stop all tests from running.
- It's better to just parametrize on the variable we actually are interested in. All the `None` make it unclear what it tested in that particular test. It also makes test hard to read in the log:
```
qutip/tests/test_visualization.py::test_hinton[rho0-None-None-scaled-True-None-None-None-None] PASSED [ 16%]
qutip/tests/test_visualization.py::test_hinton[rho1-None-None-scaled-True-None-None-None-None] PASSED [ 16%]
qutip/tests/test_visualization.py::test_hinton[rho2-None-None-scaled-True-None-None-None-None] PASSED [ 16%]
qutip/tests/test_visualization.py::test_hinton[rho3-None-None-scaled-True-None-None-None-None] PASSED [ 16%]
qutip/tests/test_visualization.py::test_hinton[rho4-x_basis4-None-scaled-True-None-None-None-None] PASSED [ 16%]
qutip/tests/test_visualization.py::test_hinton[rho5-None-y_basis5-scaled-True-None-None-None-None] PASSED [ 16%]
qutip/tests/test_visualization.py::test_hinton[rho6-None-None-scaled-True-None-None-None-None] PASSED [ 16%]
qutip/tests/test_visualization.py::test_hinton[rho7-None-None-threshold-True-None-None-None-None] PASSED [ 16%]
qutip/tests/test_visualization.py::test_hinton[rho8-None-None-phase-True-None-None-None-None] PASSED [ 16%]
qutip/tests/test_visualization.py::test_hinton_ValueError[rho0-None-None-scaled-True-None-None-None-None-Input quantum object must be an operator or superoperator.] PASSED [ 16%]
qutip/tests/test_visualization.py::test_hinton_ValueError[rho1-None-None-color_style-True-None-None-None-None-Unknown color style color_style for Hinton diagrams.] PASSED [ 16%]
qutip/tests/test_visualization.py::test_hinton_ValueError[rho2-None-None-scaled-True-None-None-None-None-Hinton plots of superoperators are currently only supported for qubits.] PASSED [ 16%]
```


```suggestion

def to_oper_bra(oper):
    return qutip.operator_to_vector(oper).dag()
    
def to_oper(oper):
    return oper

@pytest.mark.parametrize('transform, args', [
    (to_oper, {}),
    (qutip.operator_to_vector(), {}),
    (to_oper_bra, {}),
    (qutip.to_super(), {}),
    (to_oper, {'x_basis': [1, 2, 3, 4, 5]),
    (to_oper, {'y_basis': [1, 2, 3, 4, 5]),
    (to_oper, {'color_style': 'threshold'),
    (to_oper, {'color_style': 'phase'),
])
def test_hinton(transform, args):
    rho = transform(qutip.rand_dm(5))
    fig, ax = qutip.hinton(rho, **args)
```

**Comment by tamakoshi2001** [comment]
@Ericgig Hi. The Github test fails because of plot_spin_distribution. ```plot_spin_distribution``` plots color at (x, y), but the correct way to use ```pcolor``` is plotting color in rectangles made by four matrix elements ([the link](https://matplotlib.org/3.1.1/api/_as_gen/matplotlib.pyplot.pcolor.html)).  It may be difficult to fix this warning and hence we can not add pytests for it.
https://matplotlib.org/3.1.1/api/_as_gen/matplotlib.pyplot.pcolor.html

**Comment by Ericgig** [comment]
You can ignore warnings with:
`@pytest.mark.filterwarnings("ignore:The input coordinates to pcolor:UserWarning")`
The `The input coordinates to pcolor` is the start of the warning message. 
`UserWarning` is the kind of warnings.

**Comment by tamakoshi2001** [comment]
@Ericgig Hi. I added pytests for functions except for plot_wigner_sphere, matrix_histogram, and matrix_histogram_complex. Do you have any advice on them? I will cover inner functions this weekends

**Comment by Ericgig** [comment]
There are a few options:
- Wait for #2193 to be merged and add the missing tests after.
- Merge this and add the tests for `matrix_histogram`, etc. in #2193.

Any is good, just tell me when you are ready for a review.

**Comment by tamakoshi2001** [comment]
@Ericgig Hi. I prefer the latter. I am adding tests in #2193 now.

**Comment by tamakoshi2001** [comment]
@Ericgig This PR is ready for review.

File: qutip/tests/test_visualization.py [hunk 2]
```diff
@@ -0,0 +1,392 @@
...
+
+@pytest.mark.parametrize('transform, args, error_message', [
+    (to_oper, {},
+     "Input quantum object must be an operator or superoperator.")
+])
+def test_hinton_ValueError0(transform, args, error_message):
+    rho = transform(qutip.basis(2, 0))
+    with pytest.raises(ValueError) as exc_info:
+        fig, ax = qutip.hinton(rho, **args)
+    assert str(exc_info.value) == error_message
```
Comment (Ericgig) [inline qutip/tests/test_visualization.py:128]: `transform` here is not needed.

**Review by Ericgig — COMMENTED** [review]
There are a few unneeded parametrization, but overall look great.

File: qutip/tests/test_visualization.py [hunk 3]
```diff
@@ -0,0 +1,392 @@
...
+
+def test_plot_qubism_mock(monkeypatch: pytest.MonkeyPatch):
+    text = "For 'pairs_skewed' pairs of dimensions need to be the same."
+
+    def mock_func(ket, kwarg):
+        return ket
+    monkeypatch.setattr("qutip.visualization.tensor", mock_func)
+    with pytest.raises(Exception) as exc_info:
+        qutip.plot_qubism(qutip.ket("010"), how='pairs_skewed')
+    assert str(exc_info.value) == text
```
[the remark this answers is not in the document]
Comment (tamakoshi2001) [reply qutip/tests/test_visualization.py]: Thank you! It looks nice.

**Pull request merged**